### PR TITLE
Added allocations to distributors

### PR DESCRIPTION
**Pull request by jshaofuturewei**
Added allocations to distributors
1. Run "hack/globalscheduler/globalscheduler-up.sh"
2. Run "cluster/kubectl.sh apply -f globalscheduler/test/yaml/sample_2_schedulers.yaml"
3. Run "cluster/kubectl.sh apply -f globalscheduler/test/yaml/sample_4_clusters.yaml"
4. Run "cluster/kubectl.sh apply -f globalscheduler/test/yaml/sample_2_distributors.yaml"
5. Run "cluster/kubectl.sh apply -f globalscheduler/test/yaml/sample_2_allocations.yaml"
6. Run "cluster/kubectl.sh get allocations" and found the 2 allocations status phases have changed to "Assigned"

### CODE REVIEW

File: globalscheduler/controllers/distributor/distributor_process.go [hunk 1]
```diff
@@ -309,3 +319,95 @@ func matchGeoLocation(geoLocation *clustercrdv1.GeolocationInfo, podLoc v1.Resou
...
+			util.CheckTime(alloc.GetName(), "distributor", "CreateAllocation-Start", 1)
+			go func() {
+				p.ScheduleAllocation(alloc)
+			}()
+		},
+	})
+	return allocInformer
+}
+
+// ScheduleOne is to process pods by assign a scheduler to it
```
Comment (pdgetrf) [inline globalscheduler/controllers/distributor/distributor_process.go:348]: ScheduleOne --> ScheduleAllocation ?

File: globalscheduler/controllers/distributor/distributor_process.go [hunk 2]
```diff
@@ -309,3 +319,95 @@ func matchGeoLocation(geoLocation *clustercrdv1.GeolocationInfo, podLoc v1.Resou
...
+				break
+			}
+		}
+	}
+	seed := len(filteredSchedulers)
+	if seed == 0 {
+		return "", errors.New("failed to find a scheduler that fits allocation")
+	}
+	return filteredSchedulers[rand.Intn(seed)].Name, nil
+}
```
Comment (pdgetrf) [inline globalscheduler/controllers/distributor/distributor_process.go:393]: need unit test for this kind of functions unless it's POC work.

Comment (pdgetrf) [reply globalscheduler/controllers/distributor/distributor_process.go:393]: same for isAllocationGeoLocationMatched

File: globalscheduler/controllers/distributor/distributor_process.go [hunk 3]
```diff
@@ -309,3 +319,95 @@ func matchGeoLocation(geoLocation *clustercrdv1.GeolocationInfo, podLoc v1.Resou
...
+			go func() {
+				p.ScheduleAllocation(alloc)
+			}()
+		},
+	})
+	return allocInformer
+}
+
+// ScheduleOne is to process pods by assign a scheduler to it
+func (p *Process) ScheduleAllocation(alloc *allocv1.Allocation) {
```
Comment (pdgetrf) [inline globalscheduler/controllers/distributor/distributor_process.go:349]: nit: alloc --> allocation, not really saving many words

**Review by pdgetrf — COMMENTED** [review]
a few comments.

I notice that we haven't really done unit testing a lot in our PRs. Any specific reason?

File: globalscheduler/controllers/distributor/distributor_process.go [hunk 4]
```diff
@@ -309,3 +319,95 @@ func matchGeoLocation(geoLocation *clustercrdv1.GeolocationInfo, podLoc v1.Resou
...
+		alloc.Status.Phase = allocv1.AllocationAssigned
+		alloc.Status.DistributorName = p.name
+		alloc.Status.SchedulerName = scheduler
+		alloc.Status.DispatcherName = ""
+		alloc.Status.ClusterNames = []string{}
+		if _, err := p.allocClientset.GlobalschedulerV1().Allocations(alloc.Namespace).Update(alloc); err != nil {
+			klog.Warningf("Failed to assign scheduler %v to allocation %v with the error %vs", scheduler, alloc, err)
+			alloc.Status.Phase = allocv1.AllocationFailed
+			if _, err = p.allocClientset.GlobalschedulerV1().Allocations(alloc.Namespace).Update(alloc); err != nil {
+				klog.Warningf("Failed to update allocation %v - %v status to failed", alloc.Namespace, alloc.Name)
```
[the remark this answers is not in the document]
Comment (jshaofuturewei) [reply globalscheduler/controllers/distributor/distributor_process.go:367]: The only difference between line 363 and line 366 is the allocation status. The others are the same. If both updating are failed, we will know there is something wrong with updating status in line 367. Not distributor name, scheduler name, dispatcher name, and cluster names cause the issue. It is a little confusing there. I will add comments.

File: globalscheduler/controllers/distributor/distributor_process.go [hunk 5]
```diff
@@ -309,3 +319,95 @@ func matchGeoLocation(geoLocation *clustercrdv1.GeolocationInfo, podLoc v1.Resou
...
+}
+
+// ScheduleOne is to process pods by assign a scheduler to it
+func (p *Process) ScheduleAllocation(alloc *allocv1.Allocation) {
+	if alloc != nil {
+		klog.V(4).Infof("Found an allocation %v-%v to schedule:", alloc.Namespace, alloc.Name)
+		scheduler, err := p.findScheduler(alloc)
+		if err != nil {
+			klog.Warningf("Failed to find scheduler that fits scheduler with the error %v", err)
+			return
```
[the remark this answers is not in the document]
Comment (jshaofuturewei) [reply globalscheduler/controllers/distributor/distributor_process.go:355]: It won't be processed util new schedulers are added, which will trigger add event to make new scheduler to process the pod again

[the remark this answers is not in the document]
Comment (jshaofuturewei) [reply globalscheduler/controllers/distributor/distributor_process.go:355]: A new scheduler is added or an existing scheduler is updated.  The code logic is to compare scheduler geoLocation and allocation/pod geoLocation to see if there are matched. If there is nothing changed, we won't retry.

[the remark this answers is not in the document]
Comment (jshaofuturewei) [reply globalscheduler/controllers/distributor/distributor_process.go:355]: When new scheduler is created or an old scheduler is updated, the local store cache will be cleaned and the pod will by synchronized by informers to trigger another add event, which causes scheduler to process that pod again if the pod status remains the same. We found it by reading source codes. It is very interesting that events are triggered in that way.

[the remark this answers is not in the document]
Comment (pdgetrf) [reply globalscheduler/controllers/distributor/distributor_process.go:355]: okay I will need to read that part a bit more. thanks for the explanation.

**Review by pdgetrf — APPROVED** [review]
minor comments about wording.

**Review by kimeunju108 — APPROVED** [review]
LGTM

**Comment by pdgetrf** [comment]
this ready for merge right?

**Comment by jshaofuturewei** [comment]
> this ready for merge right?

Yes.

**Pull request merged**